File: source/codes/leetcode/src/main/java/leetcode/L_152.go
package sample

func maxProduct(nums []int) int {
	if len(nums) < 1 {
		return 0
	}

	maxF := make([]int, len(nums))
	minF := make([]int, len(nums))
	maxF[0] = nums[0]
	minF[0] = nums[0]
	for i := 1; i < len(nums); i++ {
		maxF[i] = max(maxF[i-1]*nums[i], max(nums[i], minF[i-1]*nums[i]))

		minF[i] = min(minF[i-1]*nums[i], min(nums[i], maxF[i-1]*nums[i]))
	}

	maxV := maxF[0]
	for i := range maxF {
		maxV = max(maxF[i], maxV)
	}
	return maxV
}

//func maxProduct(nums []int) int {
//	maxF, minF, ans := nums[0], nums[0], nums[0]
//	for i := 1; i < len(nums); i++ {
//		mx, mn := maxF, minF
//		maxF = max(mx*nums[i], max(nums[i], mn*nums[i]))
//		minF = min(mn*nums[i], min(nums[i], mx*nums[i]))
//		ans = max(maxF, ans)
//	}
//	return ans
//}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

File: assets/codes/leetcode/src/main/java/top100liked/L_152.java
package top100liked;

public class L_152 {

    public int maxProduct(int[] nums) {
        int[] maxProduct = new int[nums.length + 1];
        int[] minProduct = new int[nums.length + 1];
        int max = nums[0];
        maxProduct[0] = nums[0];

        for (int i = 1; i < nums.length; i++) {
            maxProduct[i] = Math.max(nums[i], Math.max(maxProduct[i - 1] * nums[i], minProduct[i - 1] * nums[i]));
            minProduct[i] = Math.min(nums[i], Math.min(maxProduct[i - 1] * nums[i], minProduct[i - 1] * nums[i]));
            max = Math.max(max, Math.max(maxProduct[i], minProduct[i]));
        }
        return max;
    }

    public static void main(String[] args) {
        L_152 l152 = new L_152();
        System.out.println(l152.maxProduct(new int[]{2, 3, -2, 4}));
        System.out.println(l152.maxProduct(new int[]{-2, 0, -1}));
        System.out.println(l152.maxProduct(new int[]{-2, -3, 1, 2}));
        System.out.println(l152.maxProduct(new int[]{-2, 3, -4, -5, -6}));
    }
}
